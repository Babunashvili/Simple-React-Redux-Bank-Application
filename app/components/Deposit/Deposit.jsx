import React, { Component, PropTypes } from 'react'
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import constants from '../../constants'
import FormGroup from 'react-bootstrap/lib/FormGroup'
import ControlLabel from 'react-bootstrap/lib/ControlLabel'
import FormControl from 'react-bootstrap/lib/FormControl'
import Button from 'react-bootstrap/lib/Button'
import Col from 'react-bootstrap/lib/Col'
import Row from 'react-bootstrap/lib/Row'
import Panel from 'react-bootstrap/lib/Panel'

import handleDeposit from '../../actions/depositAction'

import { checkEmptyValue, checkAmountQty, checkCardBalance } from '../../services/validation'

class Deposit extends Component {
	constructor(props) {
		super(props)
		this.state = {
			value: '',
			card:'',
			depositObject: {}
		}
	}

	/**
	 * Handle Deposit Input Changes
	 */
	onChangeHandle(e){
		this.setState({value: e.target.value, permission: true})
		
	}
	/**
	 * Handle Deposit Form Submit
	 */
	getAmount(e){
		e.preventDefault()
		if(checkEmptyValue(this.state.value)){
			//If Deposit Amount Is Not Empty
			if(checkEmptyValue(this.state.card)){
	            //If Credit Card Is Selected
				if(checkCardBalance(this.state.value,this.state.card)){
				  //If Requested Amount <= Credit Card Balance
                  if(checkAmountQty(this.state.value)){
						//If Deposit Amount > 0
						
		                this.props.handleDeposit(this.state.value, this.state.card)
						this.setState({
							value: ''
						}) 
						// this.props.handleAlert(constants.ALERT.SUCCESS_DEPOSIT_MSG,'success')
					}else{
						//If Deposit Amount <= 0
						this.props.handleAlert(constants.ALERT.NULL_DEPOSIT_MSG,'danger')
					}
				}else{
                   //If Is Not Enough Money On Card
					this.props.handleAlert(constants.ALERT.NOT_ENOUGH_BALANCE_ON_CARD_MSG,'danger')
				}
			}else{
               this.props.handleAlert(constants.ALERT.CARD_NOT_SELECTED,'danger')
			}
		}else{
			//If Deposit Amount Is Empty
			this.props.handleAlert(constants.ALERT.EMPTY_DEPOSIT_MSG,'danger') 
		}
	}
    /**
	 * Handle Deposit Card Change
	 */
	handleCardChange(e){
        this.setState({
        	card:e.target.value
        })
	}

	render() {
		return (
			<div>
				<Panel header="Deposit into your Balance">
			      <form>
					<Row>
	                     <Col md={6}>
	                       <FormGroup>
							<ControlLabel> Deposit Amount: </ControlLabel>
							<FormControl
								type="number"
								placeholder="Enter amount"
								value={this.state.value}
								onChange={this.onChangeHandle.bind(this)}
							/>
						  </FormGroup>
						</Col>
						<Col md={6}>
	                       <FormGroup>
							<ControlLabel> Choose Credit Card: </ControlLabel>
							<FormControl componentClass="select" placeholder="Choose Credit Card" onChange={this.handleCardChange.bind(this)}>
						        <option value=""></option>
						        {this.props.cards.map((value,key) => {
						        	return <option key={key} value={value.key}>{value.card.number} ({value.balance} USD)</option>
						        })}
						     </FormControl>
						  </FormGroup>
						</Col>
					</Row>
					<FormGroup>
						<Button onClick={this.getAmount.bind(this)} type="submit" bsStyle="default">Submit</Button>
					</FormGroup>
				 </form>
			    </Panel>
			</div>
		)
	}
}

/**
 * Add Deposit Component PropTypes
 */
Deposit.propTypes = {
	handleDeposit: React.PropTypes.func,
	handleAlert: React.PropTypes.func
}


const stateProps = (state) => {
	return {
		cards: state.deposit.cards
	}
}

const dispatchtToProps = (dispatch) => {
	return bindActionCreators({
		handleDeposit: handleDeposit
	}, dispatch)
} 

export default connect(stateProps, dispatchtToProps)(Deposit)