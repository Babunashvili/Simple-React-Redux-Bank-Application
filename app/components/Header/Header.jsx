import React, { Component } from "react"

import Navbar from 'react-bootstrap/lib/Navbar'
import Nav from 'react-bootstrap/lib/Nav'
import FormControl from 'react-bootstrap/lib/FormControl'
import Button from 'react-bootstrap/lib/Button'

class Header extends Component {
	render() {
		return (

			<Navbar>
				<Navbar.Header>
					<Navbar.Brand>
						<a href="#">Bank App</a>
					</Navbar.Brand>
				</Navbar.Header>
				<Navbar.Text pullRight>
					Balance <span>0</span> $
				</Navbar.Text>
				
			</Navbar>
			
		)
	}
}


export default Header