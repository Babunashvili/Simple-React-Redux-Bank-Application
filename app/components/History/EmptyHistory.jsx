import React, { Component, PropTypes } from 'react'

class EmptyHistory extends Component {
	constructor(props) {
		super(props)
	}
	render() {
		return (
			<div>
				{this.props.message}
			</div>
		)
	}
}
EmptyHistory.propTypes = {
	message: React.PropTypes.string,
}
export default EmptyHistory