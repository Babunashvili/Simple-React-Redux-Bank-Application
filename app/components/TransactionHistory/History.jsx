import React, { Component } from "react"

import Table from 'react-bootstrap/lib/Table'


class History extends Component {
	constructor(props) {
		super(props)
		
	}
	render() {
		console.log(this.props.trans)
		return (
			<div className="well">
				<h3 className="text-center">Transactions History</h3>
				<hr/>
				<Table responsive>
					<thead>
						<tr>
							<th>Date</th>
							<th>Description</th>
							<th>Amount</th>
						</tr>
					</thead>
					<tbody>
						{this.props.trans.map((value, key) => {

							return <tr key={key}>
								<td>{value.date}</td>
								<td>{value.description}</td>
								<td>{value.amount}</td>
							</tr>
						})}

					</tbody>
				</Table>
			</div>
		)
	}
}


export default History