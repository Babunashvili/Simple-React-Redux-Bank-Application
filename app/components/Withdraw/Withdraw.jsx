import React, { Component, PropTypes } from "react"

import FormGroup from 'react-bootstrap/lib/FormGroup'
import ControlLabel from 'react-bootstrap/lib/ControlLabel'
import FormControl from 'react-bootstrap/lib/FormControl'
import Button from 'react-bootstrap/lib/Button'
import Panel from 'react-bootstrap/lib/Panel'

import { checkBalance } from '../../services/validation'
import { checkEmptyAmount } from '../../services/validation'
import { checkAmountQty } from '../../services/validation'

class Withdraw extends Component {

	constructor(props) {
		super(props)
		this.state = {
			value: ''
		}
	}
	onChangeHandle(e){
		this.setState({value: e.target.value, permission: true})
	}


	getAmount(e){
		e.preventDefault()
		if(checkEmptyAmount(this.state.value)){
			if(checkAmountQty(this.state.value)){
                if(checkBalance(this.state.value,this.props.balance)){
	                this.props.handleWithdraw(this.state.value, this.state.permission)
					this.setState({
						value: ''
					})
			    }else{
				  console.log('You dont have enough money for withdraw!')
		     	}
			}else{
				console.log('Withdrow amount must be > 0')
			}
		}else{
			console.log('Withdrow Amount is required!') 
		}
		 
	}

	render() {
		return (
			<div>
			    <Panel header="Withdraw From Your Balance">
			      <form>
						<FormGroup>
							<ControlLabel> Withdraw Amount: </ControlLabel>
							<FormControl
								type="number"
								placeholder="Enter amount"
								value={this.state.value}
								onChange={this.onChangeHandle.bind(this)}
							/>
						</FormGroup>
						<FormGroup>
							<Button onClick={this.getAmount.bind(this)} type="submit" bsStyle="default">Submit</Button>
						</FormGroup>
					</form>
			    </Panel>
			</div>
		)
	}
}

Withdraw.propTypes = {
	handleWithdraw: React.PropTypes.func
}

export default Withdraw