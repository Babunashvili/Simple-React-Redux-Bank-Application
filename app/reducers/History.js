import constants from '../constants'


const InitialState = {
    balance: 0,
    transactions: [],
    cards:[
     
    ]
}
/**
 * Create Transaction Reducer
 *
 * @class      Transaction (state,action)
 * @param      {OBject}  state   The state
 * @param      {Object}  action  The action
 * @return     {Object}  { returns updated state object }
 */
  
const Transaction = (state = InitialState, action) => {

    switch (action.type) {
        case constants.ON_TRANSACTION:
            console.log(action)
            //If Action Is Transaction
            let array = [...state.transactions]
            array.push(action.payload)
            console.log(array)
            return Object.assign({}, state, {
                transactions: array
            })
        case constants.FETCH_DATA:
            return Object.assign({}, state, action.payload)
        default:
            return state
    }

}

export default Transaction