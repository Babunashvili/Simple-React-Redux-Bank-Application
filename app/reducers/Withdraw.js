import constants from '../constants'

const InitialState = {
    balance: 0,
    transactions: [],
    cards:[
      { 
        key:'343433434',
        balance:100,
        card: {number:'1234-5678-2345-7890',expires:'05.12.2017',cvc:'123'}
      },
      { 
        key:'466433546',
        balance:1600,
        card: {number:'2456-2246-9524-2252',expires:'08.11.2017',cvc:'785'}
      }
    ]
}

const Withdraw = (state = InitialState, action) => {

    switch (action.type) {
        case constants.WITHDRAW_FROM_ACCOUNT:
            //If Action Is Withdraw Request
            return Object.assign({}, state, {
                balance: state.balance - parseFloat(action.amount)
            })
        default:
            return state
    }

}



export default Withdraw