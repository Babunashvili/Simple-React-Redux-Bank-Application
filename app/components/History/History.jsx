import React, { Component, PropTypes } from 'react'
import { connect } from 'react-redux'
import Table from 'react-bootstrap/lib/Table'
import Panel from 'react-bootstrap/lib/Panel'
import HistoryTable from './HistoryTable'
import EmptyHistory from './EmptyHistory'
import constants from '../../constants'

class History extends Component {
	constructor(props) {
		super(props)
	}
    /**
     * Check transactions history
     */
	HistoryContent(){
			if(this.props.trans.length > 0)
				return <HistoryTable trans={this.props.trans} />
			else 
                return <EmptyHistory message={constants.ALERT.NO_TRANSACTIONS_HISTORY} />
	}

	render() {
		return (
			<div>
				<Panel header="Your Transactions History">
			      {this.HistoryContent()}
			    </Panel>
			</div>
		)
	}
}
/**
 * Add History Component PropTypes
 */
History.propTypes = {
	trans: React.PropTypes.arrayOf(React.PropTypes.shape({
     date: React.PropTypes.string,
     amount: React.PropTypes.string,
     description: React.PropTypes.string  
   }))
}

const stateProps = (state) => {
	return {
		trans: state.transaction.transactions
	}
}

export default connect(stateProps)(History)