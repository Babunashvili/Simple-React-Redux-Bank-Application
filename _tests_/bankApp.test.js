import React from 'react'
import renderer from 'react-test-renderer'

import Deposit from '../app/components/Deposit/Deposit'
import Withdraw from '../app/components/Withdraw/Withdraw'
import Header from '../app/components/Header/Header'
import History from '../app/components/History/History'


describe('check components if Match', () => {
	it('Deposit component Checking', () => {
		const tree = renderer.create(
			<Deposit />
		).toJSON()

		expect(tree).toMatchSnapshot()

	})
	it('Withdraw component Checking', () => {
		const tree = renderer.create(
			<Withdraw />
		).toJSON()

		expect(tree).toMatchSnapshot()

	})
	it('Header component Checking', () => {
		const tree = renderer.create(
			<Header />
		).toJSON()

		expect(tree).toMatchSnapshot()

	})
	it('History component Checking', () => {
		const tree = renderer.create(
			<History />
		).toJSON()

		expect(tree).toMatchSnapshot()

	})
})