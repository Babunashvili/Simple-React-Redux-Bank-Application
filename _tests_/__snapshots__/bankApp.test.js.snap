exports[`check components if Match Deposit component Checking 1`] = `
<div>
  <div
    className="panel panel-default"
    id={undefined}>
    <div
      className="panel-heading">
      Deposit into your Balance
    </div>
    <div
      className="panel-body">
      <form>
        <div
          className="form-group">
          <label
            className="control-label"
            htmlFor={undefined}>
             Deposit Amount: 
          </label>
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Enter amount"
            type="number"
            value="" />
        </div>
        <div
          className="form-group">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="submit">
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`check components if Match Header component Checking 1`] = `
<nav
  className="navbar navbar-default">
  <div
    className="container">
    <div
      className="navbar-header">
      <a
        className="navbar-brand"
        href="#">
        Simple Bank Application
      </a>
    </div>
    <p
      className="navbar-text navbar-right">
      Your balance is 
      <span
        className="label label-primary">
         USD
      </span>
          
    </p>
  </div>
</nav>
`;

exports[`check components if Match Withdraw component Checking 1`] = `
<div>
  <div
    className="panel panel-default"
    id={undefined}>
    <div
      className="panel-heading">
      Withdraw From Your Balance
    </div>
    <div
      className="panel-body">
      <form>
        <div
          className="form-group">
          <label
            className="control-label"
            htmlFor={undefined}>
             Withdraw Amount: 
          </label>
          <input
            className="form-control"
            id={undefined}
            onChange={[Function]}
            placeholder="Enter amount"
            type="number"
            value="" />
        </div>
        <div
          className="form-group">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="submit">
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
